code/modules/mob/mob_helpers.dm

```diff
@@ -51,134 +51,127 @@
 /**
   * Convert random parts of a passed in message to stars
   *
-  * * n - the string to convert
-  * * pr - probability any character gets changed
+  * * phrase - the string to convert
+  * * probability - probability any character gets changed
   *
   * This proc is dangerously laggy, avoid it or die
   */
-/proc/stars(n, pr)
-	n = html_encode(n)
-	if (pr == null)
-		pr = 25
-	if (pr <= 0)
-		return null
-	else
-		if (pr >= 100)
-			return n
-	var/te = n
-	var/t = ""
-	n = length(n)
-
-	for(var/p = 1 to min(n,MAX_BROADCAST_LEN))
-		if ((copytext(te, p, p + 1) == " " || prob(pr)))
-			t = text("[][]", t, copytext(te, p, p + 1))
+/proc/stars(phrase, probability = 25)
+	if(probability <= 0)
+		return phrase
+	phrase = html_decode(phrase)
+	var/leng = length(phrase)
+	. = ""
+	var/char = ""
+	for(var/i = 1, i <= leng, i += length(char))
+		char = phrase[i]
+		if(char == " " || !prob(probability))
+			. += char
 		else
-			t = text("[]*", t)
-	if(n > MAX_BROADCAST_LEN)
-		t += "..." //signals missing text
-	return sanitize(t)
+			. += "*"
+	return sanitize(.)
+

```

Comment:
not convinced this is going to work considering that index gets bytes

Reply:
You're gonna have to be more specific

Reply:
http://www.byond.com/forum/post/2520672#comment25078860
> To handle these special cases, BYOND has introduced _char versions of most of the text procs, like findtext_char() and copytext_char(). Here is one way you could, but shouldn't, update the previous code:
> 
> ```
> // This works in 513+ but performance sucks for long strings
> proc/AnimateMaptext(object, text)
>     var/i, len=length_char(text)
>     for(i=2, i<=len, ++i)
>         object.maptext = copytext_char(text, 1, i)
>         sleep(1)
> 
> ```
> Counting characters is not necessarily a fast process, so for long text strings this would be cumbersome. Here's a better way:
> 
> ```
> // This is better (and will also work in 512!)
> proc/AnimateMaptext(object, text)
>     var/i, len=length(text)
>     var/ch
>     for(i=1, i<=len,)
>         // the read-only [] operator will grab the whole character at byte i
>         i += length(text[i])
>         object.maptext = copytext(text, 1, i)
>         sleep(1)
> ```

Byond/lummox suggests that index reads are a better solution.

Reply:
The key piece you might be missing is that the length of char is added to i, ensuring it is aligned to the start of the next character.